crates/eww/src/widgets/systray.rs

```diff
@@ -0,0 +1,175 @@
+use notifier_host::{self, export::ordered_stream::OrderedStreamExt};
+
+// DBus state shared between systray instances, to avoid creating too many connections etc.
+struct DBusGlobalState {
+    con: zbus::Connection,
+    name: zbus::names::WellKnownName<'static>,
+    snw: notifier_host::dbus::StatusNotifierWatcherProxy<'static>,
+}
+
+async fn dbus_state() -> std::sync::Arc<DBusGlobalState> {
+    use once_cell::sync::Lazy;
+    use std::sync::{Arc, Weak};
+    use tokio::sync::Mutex;
+    static DBUS_STATE: Lazy<Mutex<Weak<DBusGlobalState>>> = Lazy::new(Default::default);
+
+    let mut dbus_state = DBUS_STATE.lock().await;
+    if let Some(state) = dbus_state.upgrade() {
+        state
+    } else {
+        // TODO error handling?
+        let con = zbus::Connection::session().await.unwrap();
+        notifier_host::Watcher::new().attach_to(&con).await.unwrap();
+
+        let name = notifier_host::attach_new_wellknown_name(&con).await.unwrap();
+        let snw = notifier_host::register_to_watcher(&con, &name).await.unwrap();
+
+        let arc = Arc::new(DBusGlobalState { con, name, snw });
+        *dbus_state = Arc::downgrade(&arc);
+
+        arc
+    }
+}
+
+pub struct Props {
+    icon_size_tx: tokio::sync::watch::Sender<i32>,
+}
+
+impl Props {
+    pub fn new() -> Self {
+        let (icon_size_tx, _) = tokio::sync::watch::channel(24);
+        Self { icon_size_tx }
+    }
+
+    pub fn icon_size(&self, value: i32) {
+        let _ = self.icon_size_tx.send_if_modified(|x| {
+            if *x == value {
+                false
+            } else {
+                *x = value;
+                true
+            }
+        });
+    }
+}
+
+struct Tray {
+    menubar: gtk::MenuBar,
+    items: std::collections::HashMap<String, Item>,
+
+    icon_size: tokio::sync::watch::Receiver<i32>,
+}
+
+pub fn spawn_systray(menubar: &gtk::MenuBar, props: &Props) {
+    let mut systray = Tray { menubar: menubar.clone(), items: Default::default(), icon_size: props.icon_size_tx.subscribe() };
+
+    // TODO when does this task die?
+    glib::MainContext::default().spawn_local(async move {
+        let s = &dbus_state().await;
+        systray.menubar.show();
+        notifier_host::run_host_forever(&mut systray, &s.snw).await.unwrap();
+    });
+}
+
+impl notifier_host::Host for Tray {
+    fn add_item(&mut self, id: &str, item: notifier_host::Item) {
+        let item = Item::new(id.to_owned(), item, self.icon_size.clone());
+        self.menubar.add(&item.mi);
+        if let Some(old_item) = self.items.insert(id.to_string(), item) {
+            self.menubar.remove(&old_item.mi);
+        }
+    }
+
+    fn remove_item(&mut self, id: &str) {
+        if let Some(item) = self.items.get(id) {
+            self.menubar.remove(&item.mi);
+        } else {
+            log::warn!("Tried to remove nonexistent item {:?} from systray", id);
+        }
+    }
+}
+
+struct Item {
+    mi: gtk::MenuItem,
+
+    tasks: Vec<glib::SourceId>,
+}
```

Comment:
This struct and its field definitely need some doc comments, it's very non-obvious what exactly this represents